# PIAIC - AIC - Assignment_1

## Data

In [1]:
# Input Data
labels = ['gender', 'parental level of education', 'test preparation course', 'math score', 'reading score', 'writing score'],
data = [
   ['female', "bachelor's degree", 'none', '72', '72', '74'],
  ['female', 'some college', 'completed', '69', '90', '88'],
  ['female', "master's degree", 'none', '90', '95', '93'],
  ['male', "associate's degree", 'none', '47', '57', '44'],
  ['male', 'some college', 'none', '76', '78', '75'],
  ['female', "associate's degree", 'none', '71', '83', '78'],
  ['female', 'some college', 'completed', '88', '95', '92'],
  ['male', 'some college', 'none', '40', '43', '39'],
  ['male', 'high school', 'completed', '64', '64', '67'],
  ['female', 'high school', 'none', '38', '60', '50'],
  ['male', "associate's degree", 'none', '58', '54', '52'],
 ['male', "associate's degree", 'none', '40', '52', '43'],
 ['female', 'high school', 'none', '65', '81', '73'],
 ['male', 'some college', 'completed', '78', '72', '70'],
 ['female', "master's degree", 'none', '50', '53', '58'],
 ['female', 'some high school', 'none', '69', '75', '78'],
 ['male', 'high school', 'none', '88', '89', '86'],
 ['female', 'some high school', 'none', '18', '32', '28'],
 ['male', "master's degree", 'completed', '46', '42', '46'],
 ['female', "associate's degree", 'none', '54', '58', '61'],
 ['male', 'high school', 'none', '66', '69', '63'],
 ['female', 'some college', 'completed', '65', '75', '70'],
 ['male', 'some college', 'none', '44', '54', '53'],
 ['female', 'some high school', 'none', '69', '73', '73'],
 ['male', "bachelor's degree", 'completed', '74', '71', '80'],
 ['male', "master's degree", 'none', '73', '74', '72'],
 ['male', 'some college', 'none', '69', '54', '55'],
 ['female', "bachelor's degree", 'none', '67', '69', '75'],
 ['male', 'high school', 'none', '70', '70', '65'],
 ['female', "master's degree", 'none', '62', '70', '75'],
 ['female', 'some college', 'none', '69', '74', '74'],
 ['female', 'some college', 'none', '63', '65', '61'],
 ['female', "master's degree", 'none', '56', '72', '65'],
 ['male', 'some college', 'none', '40', '42', '38'],
 ['male', 'some college', 'none', '97', '87', '82'],
 ['male', "associate's degree", 'completed', '81', '81', '79'],
 ['female', "associate's degree", 'none', '74', '81', '83'],
 ['female', 'some high school', 'none', '50', '64', '59'],
 ['female', "associate's degree", 'completed', '75', '90', '88'],
]

## Question # 1:


In [2]:
def query(gender='male',subject='math'):
    male_scores = []
    fem_scores = []
    if gender == 'male':
        for item in data:
            if item[0] == 'male':
                if subject == 'math':
                    male_scores.append(int(item[-3]))
                elif subject == 'reading':
                    male_scores.append(int(item[-2]))
                elif subject == 'writing':
                    male_scores.append(int(item[-1]))
                else:
                    print("Wrong input! Use the Legend as reference and type subject accordingly")                                 
        return male_scores


    elif gender == 'female':
        for item in data:
            if item[0] == 'female':
                if subject == 'math':
                    fem_scores.append(int(item[-3]))
                elif subject == 'reading':
                    fem_scores.append(int(item[-2]))
                elif subject == 'writing':
                    fem_scores.append(int(item[-1]))
                else:
                    print("Wrong input! Restart & Use the Legend as reference and type 'Subject' accordingly")                                
        return fem_scores
    else:
        print("Wrong input! Restart & Use the Legend as reference and type 'Gender' accordingly")    

#### Legend for Reference:
(Note: Type these exactly as given here in the input boxes.)
* Gender: `male` or `female`

* Education: `some highschool` or `some college` or `associate's degree` or `bachelor's degree` or `master's degree`

* Subjects: `maths` or `reading` or `writing`



In [3]:
#Female Scores
fem_math=sum(query('female','math'))
fem_read=sum(query('female','reading'))
fem_write=sum(query('female','writing'))
fem_total= fem_math + fem_read + fem_write

print(f"Females scored: \n {fem_math} in 'Maths' \n {fem_read} in 'Reading'\n {fem_write} in 'Writing' \n The total score of All Females in All Subjects is {fem_total} ")


Females scored: 
 1334 in 'Maths' 
 1527 in 'Reading'
 1496 in 'Writing' 
 The total score of All Females in All Subjects is 4357 


In [4]:
#Male Scores
male_math=sum(query('male', 'math'))
male_read=sum(query('male', 'reading'))
male_write=sum(query('male','writing'))
male_total= male_math + male_read + male_write

print(f"Males scored: \n {male_math} in 'Maths' \n {male_read} in 'Reading'\n {male_write} in 'Writing' \n The total score of All Males in All Subjects is {male_total} ")

Males scored: 
 1151 in 'Maths' 
 1153 in 'Reading'
 1109 in 'Writing' 
 The total score of All Males in All Subjects is 3413 


In [6]:
#Input Function Form:
print(query(gender=input('Type male or female: '), subject=input('Type subject:')))


Type male or female: male
Type subject:math
[47, 76, 40, 64, 58, 40, 78, 88, 46, 66, 44, 74, 73, 69, 70, 40, 97, 81]


In [14]:
# Average Score Comparison
fem_avg = fem_total/(len(list(i for i in data if i[0]=='female' )))
male_avg = male_total/(len(list(i for i in data if i[0]=='male' )))

print(f"Female Average Score is '{fem_avg}', while Male Average Score is '{male_avg}'")

Female Average Score is '207.47619047619048', while Male Average Score is '189.61111111111111'


__Conclusion__:
* In __total__, males have scored **3413** in all subjects, and females have **4357**. 

* On __Average__, males have scored **189.61** in all subjects, and females have scored **207.47**

* Females have higher **Total** & **Average** Scores than males.

## Question # 2:
Is there any difference in student score if parents have __master-level education__ in a specified score (we will take user input for choosing the score column)?

In [62]:
def diff(education="some college", subject='math'):
    subj_scores = []
    for item in data:
        if item[1] == education:
            if subject == 'math':
                subj_scores.append(int(item[-3]))
            elif subject == 'reading':
                subj_scores.append(int(item[-2]))
            elif subject == 'writing':
                subj_scores.append(int(item[-1]))
    try:
        avg_score = sum(subj_scores)/len(subj_scores)
        return (f"There are '{len(subj_scores)}' students whose parents have '{education}'. \nThe average score in '{subject}' for students with parents having '{education}' is '{avg_score}'. \nIndividual Score list {subj_scores}")
    except ZeroDivisionError:
        return (f"There are '{len(subj_scores)}' students whose parents have '{education}'.")
    

#### Legend for Reference:
(Note: Type these exactly as given here in the input boxes.)
* Gender: `male` or `female`

* Education: `some highschool` or `some college` or `associate's degree` or `bachelor's degree` or `master's degree`

* Subjects: `maths` or `reading` or `writing`



In [52]:
# In Subject of "Maths"
print(diff(education=input('Type Level of Education: '), subject=input('Type Subject: ')))

Type Level of Education: master's degree
Type Subject: math
There are '6' students whose parents have 'master's degree'. 
The average score in 'math' for students with parents having 'master's degree' is '62.833333333333336'. 
Individual Score list [90, 50, 46, 73, 62, 56]


In [53]:
# In Subject of "Reading"
print(diff(education=input('Type Level of Education: '), subject=input('Type Subject: ')))

Type Level of Education: master's degree
Type Subject: reading
There are '6' students whose parents have 'master's degree'. 
The average score in 'reading' for students with parents having 'master's degree' is '67.66666666666667'. 
Individual Score list [95, 53, 42, 74, 70, 72]


In [54]:
# In Subject of "Writing"
print(diff(education=input('Type Level of Education: '), subject=input('Type Subject: ')))

Type Level of Education: master's degree
Type Subject: writing
There are '6' students whose parents have 'master's degree'. 
The average score in 'writing' for students with parents having 'master's degree' is '68.16666666666667'. 
Individual Score list [93, 58, 46, 72, 75, 65]


In [64]:
# COMPARISON WITH: [In Subject of "Writing" with "some high school"}
print(diff(education=input('Type Level of Education: '), subject=input('Type Subject: ')))

Type Level of Education: some high school
Type Subject: writing
There are '4' students whose parents have 'some high school'. 
The average score in 'writing' for students with parents having 'some high school' is '59.5'. 
Individual Score list [78, 28, 73, 59]


* __Conclusion__: On Average, In the subject of __"Writing"__ Students with parents having __master's degree__ have more score **(68.16)**, compared to the students with parents having __some high school__ degree **(59.5)**

* This function can be used to compare any subject with any parental education level

## Question # 3:
Is there any difference in student score if parents have __Bachelor level education__ in a specified score (we will take user input for choosing the score column)?

#### Legend for Reference:
(Note: Type these exactly as given here in the input boxes.)

* Education: `some highschool` or `some college` or `associate's degree` or `bachelor's degree` or `master's degree`

* Subjects: `maths` or `reading` or `writing`



In [63]:
# We can use the same function as described in the last question
# In Subject of "Maths"
print(diff(education=input('Type Level of Education: '), subject=input('Type Subject: ')))

Type Level of Education: bachelor's degree
Type Subject: maths
There are '0' students whose parents have 'bachelor's degree'.


In [65]:
# In Subject of "Reading"
print(diff(education=input('Type Level of Education: '), subject=input('Type Subject: ')))

Type Level of Education: bachelor's degree
Type Subject: reading
There are '3' students whose parents have 'bachelor's degree'. 
The average score in 'reading' for students with parents having 'bachelor's degree' is '70.66666666666667'. 
Individual Score list [72, 71, 69]


In [66]:
# In Subject of "Writing"
print(diff(education=input('Type Level of Education: '), subject=input('Type Subject: ')))

Type Level of Education: bachelor's degree
Type Subject: writing
There are '3' students whose parents have 'bachelor's degree'. 
The average score in 'writing' for students with parents having 'bachelor's degree' is '76.33333333333333'. 
Individual Score list [74, 80, 75]


* __Conclusion__: On Average, In the subject of __"Reading"__ Students with parents having __bachlor's degree__ have more score **(70.66)**, compared to the students with parents having __master's__ degree above **(67.6)**

## Question # 4:

How many parents have bachelors education, master education, or some college degrees level education?

In [2]:
def edu(list):
    bach = 0
    col = 0
    mast = 0
    asso = 0
    hs = 0
    
    for item in list:
        if item[1] == "associate's degree":
            asso += 1
        elif item[1] == 'high school':
            hs += 1
        elif item[1] == "master's degree":
            col += 1
        elif  item[1] == 'some college':
            mast += 1
        elif item[1] == "bachelor's degree":
            bach += 1
    print(f"{bach} parents have bachelor's degree \n{col} parents some college degree \n{mast} parents have master's degree \n{asso} parents have associate's degree \n{hs} parents have a high school degree")


In [3]:
edu(data)

3 parents have bachelor's degree 
6 parents some college degree 
11 parents have master's degree 
8 parents have associate's degree 
6 parents have a high school degree


## Question # 5:
In a user given score, is the top scorer a female or male?

In [4]:
def top(subject):
    male_top = 0
    female_top = 0
    
    for item in data:
        if item[0] == 'female':
            if subject == 'math':
                if female_top < int(item[-3]):
                    female_top = int(item[-3])
                else:
                    continue
            elif subject == 'reading':
                if female_top < int(item[-2]):
                    female_top = int(item[-2])
                else:
                    continue
            elif subject == 'writing':
                if female_top < int(item[-1]):
                    female_top = int(item[-1])
                else:
                    continue
        if item[0] == 'male':
            if subject == 'math':
                if male_top < int(item[-3]):
                    male_top = int(item[-3])
            elif subject == 'reading':
                if male_top < int(item[-2]):
                    male_top = int(item[-2])
                else:
                    continue
            elif subject == 'writing':
                if male_top < int(item[-1]):
                    male_top = int(item[-1])
                    
    return (f"Top score in the subject of '{subject}' for Males is {male_top}, and for Females is {female_top}")
            

In [5]:
top('math')

"Top score in the subject of 'math' for Males is 97, and for Females is 90"

In [6]:
top('reading')

"Top score in the subject of 'reading' for Males is 89, and for Females is 95"

In [7]:
top('writing')

"Top score in the subject of 'writing' for Males is 86, and for Females is 93"

In [8]:
top(input("Please enter Subject('math'/'reading'/'writing'): "))

Please enter Subject('math'/'reading'/'writing'): math


"Top score in the subject of 'math' for Males is 97, and for Females is 90"

## Question 6:
How many students have good in reading (> 75) but not good in writing( < 70)?

In [9]:
data[0:5]

[['female', "bachelor's degree", 'none', '72', '72', '74'],
 ['female', 'some college', 'completed', '69', '90', '88'],
 ['female', "master's degree", 'none', '90', '95', '93'],
 ['male', "associate's degree", 'none', '47', '57', '44'],
 ['male', 'some college', 'none', '76', '78', '75']]

In [10]:
def goodInRead(list):
    result = 0
    for item in list:
        if int(item[-2]) > 75 and int(item[-1]) < 70:
            result += 1
    return (f" {result} students have a good score in 'Reading' but low scores in 'Writing'")
    
    

In [11]:
goodInRead(data)

" 0 students have a good score in 'Reading' but low scores in 'Writing'"